calyx-opt/src/passes/top_down_compile_control.rs

```diff
@@ -388,195 +475,178 @@ impl<'b, 'a> Schedule<'b, 'a> {
         states.push(FSMStateInfo {
-            id: final_state,
-            group: Id::new(format!("{}_END", fsm.borrow().name())),
+            id: fsm_rep.last_state, // check that this register (fsm.0) is the correct one to use
+            group: Id::new(format!("{}_END", fsm1.borrow().name())),
         });
 
         // Keep track of groups to FSM state id information for dumping to json
         fsm_groups.insert(ProfilingInfo::Fsm(FSMInfo {
             component: self.builder.component.name,
-            fsm: fsm.borrow().name(),
+            fsm: fsm1.borrow().name(),
             group: group.borrow().name(),
             states,
         }));
 
-        // keep track of used slicers if using one hot encoding
-        let mut used_slicers = HashMap::new();
+        // keep track of used slicers if using one hot encoding. one for each register
+        let mut used_slicers_vec =
+            fsms.iter().map(|_| HashMap::new()).collect_vec();
 
         // enable assignments
+        // the following enable queries; we can decide which register to query for state-dependent assignments
+        // because we know all registers precisely agree at each cycle
         group.borrow_mut().assignments.extend(
             self.enables
                 .into_iter()
                 .sorted_by(|(k1, _), (k2, _)| k1.cmp(k2))
-                .flat_map(|(state, mut assigns)| match encoding {
-                    Encoding::Binary => {
-                        let state_const =
-                            self.builder.add_constant(state, fsm_size);
-                        let state_guard =
-                            guard!(fsm["out"] == state_const["out"]);
-                        assigns.iter_mut().for_each(|asgn| {
-                            asgn.guard.update(|g| g.and(state_guard.clone()))
-                        });
-                        assigns
-                    }
-                    Encoding::OneHot => {
-                        let state_guard = Self::build_one_hot_query(
-                            self.builder,
-                            &mut used_slicers,
-                            &fsm,
-                            &signal_on,
-                            &state,
-                            &fsm_size,
-                        );
-                        assigns.iter_mut().for_each(|asgn| {
-                            asgn.guard.update(|g| g.and(state_guard.clone()))
-                        });
-                        assigns
-                    }
+                .flat_map(|(state, mut assigns)| {
+                    // find the register from which to query; try to split evenly among registers
+                    let (fsm, used_slicers) = {
+                        let num_registers: u64 =
+                            fsms.len().try_into().unwrap();
+                        let reg_to_query : usize = (state * num_registers
+                            / (fsm_rep.last_state)).try_into().unwrap();
+
```

Comment:
Just double checking: this math is basically trying to find an even split among the regsiter queries?